### PR TITLE
[Fix #190] Add httpjson tags support

### LINKED ISSUE #190: httpjson plugin should allow JSON to add tags
The httpjson plugin should have a way for the data emitter/end-point to specify tags to add to each key/value pair. This would make it possible to make "InfluxDB-aware" (telegraf-aware, I guess to be exact) applications and just use this plugin, rather than having to make a specific plugin for each application.

### PR DESCRIPTION
**Pull request by palkan**
Tags are extracted from JSON response according to the list of tags in config.

Example:

```toml
# config
[httpjson.services]
  tagKeys = [
    "role",
    "build"
  ]
```
```json
# server response
{
  "value": 123,
  "role": "master",
  "build": "2015"
}
```

And resulting point will contain values `value: 123` and tags `role: master  build: 2015  server_name: ...`.

### CODE REVIEW

**Comment by sparrc** [comment]
awesome, thank you @palkan! I see that you have another PR to fix the test issue too :-)

Could you also write a short README for this plugin, and include the info you wrote above about tagging? You can use the [exec plugin](https://github.com/influxdb/telegraf/tree/master/plugins/exec) as an example.

**Comment by sparrc** [comment]
also, just to clarify, do these tag keys need to be in the top-level of the map?

File: plugins/httpjson/httpjson.go [hunk 1]
```diff
@@ -135,6 +142,14 @@ func (h *HttpJson) gatherServer(acc plugins.Accumulator, service Service, server
 		"server": serverURL,
 	}
 
+	for _, tag := range service.Tags {
+		switch v := jsonOut[tag].(type) {
+			case string:
+		    tags[tag] = v
```
Comment (sparrc) [inline plugins/httpjson/httpjson.go]: indentation, should all be tabs in Go

**Comment by sparrc** [comment]
I like this, and I also opened #277 to allow for a more general `tags` key to be used to add ad-hoc tags.

**Comment by sparrc** [comment]
@palkan one more thing: can you change the name `tags` to `tagkeys` so that it's more clear that we are referring to keys from the JSON that will be used as tags

File: plugins/httpjson/httpjson.go [hunk 2]
```diff
@@ -61,6 +62,12 @@ var sampleConfig = `
     # HTTP method to use (case-sensitive)
     method = "GET"
 
+    # List of tag names to extract from server response
+    tags = [
```
Comment (sparrc) [inline plugins/httpjson/httpjson.go]: comment this by default

File: plugins/httpjson/httpjson.go [hunk 3]
```diff
@@ -126,7 +133,7 @@ func (h *HttpJson) gatherServer(acc plugins.Accumulator, service Service, server
 		return err
 	}
 
-	var jsonOut interface{}
+	var jsonOut map[string]interface{}
```
Comment (sparrc) [inline plugins/httpjson/httpjson.go]: Won't this break the entire plugin if there is a non-string key in the top-level of the JSON?

Comment (palkan) [inline plugins/httpjson/httpjson.go]: Yes, response can contain JSON array. We should have a test case for this)

Comment (palkan) [inline plugins/httpjson/httpjson.go]: Hm, this plugin wasn't aware of array JSON. I think this should go to another PR (along with support for string values, do you know why the plugin supports only floats, btw?).

Comment (sparrc) [inline plugins/httpjson/httpjson.go]: @palkan supporting floats could become very complicated, there tends to be a lot of metadata in JSON that users probably don't want. Also storing strings as InfluxDB values has bad compression and isn't very useful.

**Comment by palkan** [comment]
Tags should be in the top level of the map. Of course, we can look for tags in lower levels, but we have to consider prefixes here. One possible solution is to flatten JSON first, and provide tags names with prefixes, e.g.:

```json
# json
{
  "key": 1,
  "some": {
    "key": 2
  },
  "none": {
     "key": 3
   },
}
```

and config

```toml
[httpjson]
  tagKeys = [
    "key",
    "some_key"
  ]

```

will produce `httpjson_service,key=1,some_key=2 ...`.

**Comment by palkan** [comment]
@sparrc 
We're waiting until we end up with something in #277?

**Comment by sparrc** [comment]
sorry for the delay @palkan, I'll get this merged this weekend :+1:
